### PR TITLE
basefw: add support for fw config query for context save

### PR DESCRIPTION
**Pull request by RanderWang**
If CONFIG_ADSP_IMR_CONTEXT_SAVE is enabled, base fw will report fw_context_save is supported to host

kernel pr: https://github.com/thesofproject/linux/pull/4680

### CODE REVIEW

File: src/audio/base_fw.c [hunk 1]
```diff
@@ -111,6 +111,11 @@ static int basefw_config(uint32_t *data_offset, char *data)
 	sche_cfg.sys_tick_source = SOF_SCHEDULE_LL_TIMER;
 	tlv_value_set(tuple, IPC4_SCHEDULER_CONFIGURATION, sizeof(sche_cfg), &sche_cfg);
 
+	if (IS_ENABLED(CONFIG_ADSP_IMR_CONTEXT_SAVE)) {
+		tuple = tlv_next(tuple);
+		tlv_value_uint32_set(tuple, IPC4_FW_CONTEXT_SAVE, 1);
+	}
```
Comment (ujfalusi) [inline src/audio/base_fw.c:117]: This has to be a negative flag, 1 if context save is not available and 0 if context save is implemented.

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: src/audio/base_fw.c (hunk 1 above)
Comment (RanderWang) [reply src/audio/base_fw.c]: sure, updated.

**Comment by RanderWang** [comment]
on kernel side,  I change the default value of fw_context_save on TGL & MTL to ture, unless fw overrides it

**Review by btian1 — APPROVED** [review]
just a suggestion in comments:
If CONFIG_ADSP_IMR_CONTEXT_SAVE is enabled, base fw will report
fw_context_save is supported to driver. Driver will assume fw doesn't
support this feature if it is not reported.

**Comment by ujfalusi** [comment]
> just a suggestion in comments: If CONFIG_ADSP_IMR_CONTEXT_SAVE is enabled, base fw will report fw_context_save is supported to driver. Driver will assume fw doesn't support this feature if it is not reported.

The assumption is that the context save is supported if IMR is supported. We need negative flagging for this.

File: src/audio/base_fw.c [hunk 2]
```diff
@@ -111,6 +111,12 @@ static int basefw_config(uint32_t *data_offset, char *data)
 	sche_cfg.sys_tick_source = SOF_SCHEDULE_LL_TIMER;
 	tlv_value_set(tuple, IPC4_SCHEDULER_CONFIGURATION, sizeof(sche_cfg), &sche_cfg);
 
+	tuple = tlv_next(tuple);
+	if (IS_ENABLED(CONFIG_ADSP_IMR_CONTEXT_SAVE))
+		tlv_value_uint32_set(tuple, IPC4_FW_CONTEXT_SAVE, 1);
+	else
+		tlv_value_uint32_set(tuple, IPC4_FW_CONTEXT_SAVE, 0);
```
Comment (lyakh) [inline src/audio/base_fw.c:118]: I think it should work to just use `tlv_value_uint32_set(tuple, IPC4_FW_CONTEXT_SAVE, IS_ENABLED(CONFIG_ADSP_IMR_CONTEXT_SAVE))`

Comment (RanderWang) [reply src/audio/base_fw.c:118]: is IS_ENABLED work ?

**Comment by RanderWang** [comment]
> > just a suggestion in comments: If CONFIG_ADSP_IMR_CONTEXT_SAVE is enabled, base fw will report fw_context_save is supported to driver. Driver will assume fw doesn't support this feature if it is not reported.
> 
> The assumption is that the context save is supported if IMR is supported. We need negative flagging for this.

@ujfalusi  how about ID now ?   We don't need negative flag sine the default value in kernel is true, not false.

File: src/include/ipc4/base_fw.h [hunk 3]
```diff
@@ -361,8 +361,10 @@ enum ipc4_fw_config_params {
 	IPC4_BUS_HARDWARE_ID = 25,
 	/* Total number of FW config parameters  */
 	IPC4_FW_CFG_PARAMS_COUNT,
+	/* FW context save on D3 entry */
+	IPC4_FW_CONTEXT_SAVE,
 	/* Max config parameter id */
-	IPC4_MAX_FW_CFG_PARAM = IPC4_FW_CFG_PARAMS_COUNT - 1,
+	IPC4_MAX_FW_CFG_PARAM = IPC4_FW_CONTEXT_SAVE - 1,
```
Comment (lgirdwood) [inline src/include/ipc4/base_fw.h:367]: @RanderWang @ujfalusi @mwasko @marcinszkudlinski to be on the safe side - can we add our new Intel reserved IDs as part of this PR and I think its going to be better if we also assign the numbers when updating this enum. i.e there is no ambiguos enum,  developers should always see
```
IPC4_SOME_ENUM = NUMBER,
```

**Review by lgirdwood — COMMENTED** [review]
Just need to align new IDs

File: src/include/ipc4/base_fw.h (hunk 3 above)
Comment (RanderWang) [reply src/include/ipc4/base_fw.h:367]: good idea!  I use 100, big enough for 10 years.

**Comment by RanderWang** [comment]
reserve 100 for this config

File: src/include/ipc4/base_fw.h (hunk 3 above)
Comment (lgirdwood) [reply src/include/ipc4/base_fw.h]: > good idea! I use 100, big enough for 10 years.

be careful, our industry has a bad experinces with quotes like these ;)

**Comment by RanderWang** [comment]
@lgirdwood  sync latest config from ref fw.

**Comment by RanderWang** [comment]
@lgirdwood  I reserved 3 id, how about it ? Thanks!

File: src/include/ipc4/base_fw.h [hunk 4]
```diff
@@ -359,6 +359,12 @@ enum ipc4_fw_config_params {
 	IPC4_TELEMETRY_BUFFER_SIZE = 24,
 	/* HW version information  */
 	IPC4_BUS_HARDWARE_ID = 25,
+	/* Total memory size available in debug window */
```
Comment (lgirdwood) [inline src/include/ipc4/base_fw.h:362]: is this now aligned with reference IDs ?

File: src/include/ipc4/base_fw.h [hunk 5]
```diff
@@ -365,6 +365,10 @@ enum ipc4_fw_config_params {
 	IPC4_KPB_RT_SINK_COUNT = 27,
 	/* Manual control of DMI L1 state */
 	IPC4_DMI_FORCE_L1_EXIT = 28,
+	/* Intel reserved */
+	IPC4_INTEL_RESRVED_29 = 29,
+	IPC4_INTEL_RESRVED_30 = 30,
+	IPC4_INTEL_RESRVED_31 = 31,
```
Comment (lgirdwood) [inline src/include/ipc4/base_fw.h:371]: If we are using 31 we can name it using its real name.

File: src/include/ipc4/base_fw.h (hunk 4 above)
Comment (RanderWang) [reply src/include/ipc4/base_fw.h:362]: yes, sync with latest ID

File: src/include/ipc4/base_fw.h (hunk 5 above)
Comment (RanderWang) [reply src/include/ipc4/base_fw.h]: sure

**Comment by RanderWang** [comment]
@lgirdwood  update the ID name

**Review by lgirdwood — APPROVED** [review]
@mwasko this PR also reserves the in-use IDs until some process is created to align.

**Comment by lgirdwood** [comment]
@ujfalusi pls review, does kernel align or does it require a similar PR ?

File: src/include/ipc4/base_fw.h [hunk 6]
```diff
@@ -365,6 +365,11 @@ enum ipc4_fw_config_params {
 	IPC4_KPB_RT_SINK_COUNT = 27,
 	/* Manual control of DMI L1 state */
 	IPC4_DMI_FORCE_L1_EXIT = 28,
+	/* Intel reserved */
+	IPC4_INTEL_RESRVED_29 = 29,
+	IPC4_INTEL_RESRVED_30 = 30,
```
Comment (kv2019i) [inline src/include/ipc4/base_fw.h:370]: This looks a bit odd, why not RESERVED? I don't get why save one 'e'

File: src/include/ipc4/base_fw.h [hunk 7]
```diff
@@ -359,6 +359,12 @@ enum ipc4_fw_config_params {
 	IPC4_TELEMETRY_BUFFER_SIZE = 24,
 	/* HW version information  */
 	IPC4_BUS_HARDWARE_ID = 25,
+	/* Total memory size available in debug window */
+	IPC4_DEBUG_WINDOW_SIZE = 26,
+	/* Number of Realtime KPB output pins */
```
Comment (kv2019i) [inline src/include/ipc4/base_fw.h:364]: Other entries in this enum have been explained with acronyms opened, so better to put "Number of realtime Key Phrase Buffer (KPB) output pins."

**Review by kv2019i — COMMENTED** [review]
A few comments on the code documentation, otherwise looks good.

File: src/include/ipc4/base_fw.h (hunk 6 above)
Comment (lgirdwood) [reply src/include/ipc4/base_fw.h:370]: Ah, yeah, lets fix the spelling.

Comment (RanderWang) [reply src/include/ipc4/base_fw.h]: thanks

File: src/include/ipc4/base_fw.h (hunk 7 above)
Comment (RanderWang) [reply src/include/ipc4/base_fw.h]: thanks

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@wszypelt @lrudyX not expecting an ID update to break, good to merge ? Thanks !

**Review by ujfalusi — APPROVED** [review]

**Review by kv2019i — APPROVED** [review]
@mwasko @mmaka1 good to go (this reserves 29 for the new cfg id)?

**Review by abonislawski — APPROVED** [review]

**Review by tmleman — APPROVED** [review]

**Comment by RanderWang** [comment]
SOFCI TEST

**Comment by RanderWang** [comment]
Will update it according to ref fw which has merged a config

**Comment by RanderWang** [comment]
@lgirdwood  @kv2019i  update PR to follow ref fw https://github.com/intel-innersource/drivers.audio.firmware.cavs-ace/pull/1533

**Comment by kv2019i** [comment]
SOFCI TEST

**Pull request merged**